From fce91bec5bef534e52f3261cc289a21a2cdb5fe3 Mon Sep 17 00:00:00 2001
From: Tao Liu <ltao@redhat.com>
Date: Sun, 11 Jul 2021 22:30:22 +0800
Subject: [PATCH 21/27] Fix the failure of reporting vmcore and vmlinux do not
 match for kernels(<2.6.11)

There is a regression issue for kernels(<2.6.11) as below:

    $ crash 2.6.9-68.9/vmcore 2.6.9-68.9/vmlinux.gz
    ...
    GNU gdb (GDB) 10.2
    ...
    crash: /var/tmp/vmlinux.gz_GLsAvX and 2.6.9-68.9/vmcore do not match!

The reason is that it needs to read out the address of linux banner with
readmem() first, and then the read_string() will be able to read the data
from linux banner. So, for the kernels(<2.6.11) case, lets still invoke
get_symbol_data() to accomplish this. See the changes:
[1] https://elixir.bootlin.com/linux/v2.6.10/source/init/version.c#L38
[2] https://elixir.bootlin.com/linux/v2.6.11/source/init/version.c#L38

Signed-off-by: Tao Liu <ltao@redhat.com>
Signed-off-by: Lianbo Jiang <lijiang@redhat.com>
---
 kernel.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/kernel.c b/kernel.c
index 6a812ac7b1bc..a559f937980e 100644
--- a/kernel.c
+++ b/kernel.c
@@ -1052,7 +1052,8 @@ verify_version(void)
 
 	if (!(sp = symbol_search("linux_banner")))
 		error(FATAL, "linux_banner symbol does not exist?\n");
-	else if ((sp->type == 'R') || (sp->type == 'r') || (sp->type == 'D') ||
+	else if ((sp->type == 'R') || (sp->type == 'r') ||
+		(THIS_KERNEL_VERSION >= LINUX(2,6,11) && sp->type == 'D') ||
 		 (machine_type("ARM") && sp->type == 'T') ||
 		 (machine_type("ARM64")))
 		linux_banner = symbol_value("linux_banner");
-- 
2.30.2

